Allow validation against specific schema version

pawiecz opened the pull request:
Depends-on: kernelci/kcidb-io#37

Closes: #78

Signed-off-by: Paweł Wieczorek <pawiecz@collabora.com>

spbnick commented:
Hi @pawiecz! Would you like to finish this up eventually, or do you want someone else to take over?

pawiecz commented:
Hi @spbnick! I should be able to finish this PR up during the weekend. I'll let you know if anything comes up and I'm not able to complete it.

pawiecz commented:
Updates in v2:
1. Removed additional argument parser
2. Moved optional positional argument to the parser previously used by `validate_main()`; `validate_main()` is [called without arguments when used as an entry point](https://packaging.python.org/specifications/entry-points/#use-for-scripts)
3. Changed schema version selection using provided snippet
4. Called `validate()` directly on the specific schema version

Whole code has been moved to `kcidb.validate_main()`

I'm not entirely sure if this is what you expected - please let me know if I misunderstood.

pawiecz commented:
> 2. Moved optional positional argument to the parser previously used by `validate_main()`; `validate_main()` is [called without arguments when used as an entry point](https://packaging.python.org/specifications/entry-points/#use-for-scripts)

It turns out adding optional positional argument `schema` to the `misc.ArgumentParser` conflicts with other tools using it or inheriting from it.

Adding an optional positional argument just to the `validate_main()` would not be passed to the entry point function. Let me know what do you think would be the best way forward.

spbnick commented on kcidb/__init__.py:
```diff
@@ -250,10 +250,17 @@ def validate_main():
     sys.excepthook = misc.log_and_print_excepthook
     description = 'kcidb-validate - Validate I/O JSON data'
     parser = misc.ArgumentParser(description=description)
```
We can simply add the argument right here, after creating the parser.  It would also be good to be clear that we need the schema **version** (and not e.g. a schema file) in the metavar, and describe what the provided schema version would be used for, in the help message. So, something like this:

```suggestion
    parser = misc.ArgumentParser(description=description)
    parser.add_argument(
        'schema_version',
        metavar="SCHEMA_VERSION",
        type=non_negative_int,
        help='Validate against the schema with the specified major version, '
        'instead of the latest one.',
        nargs='?'
    )
```

An improvement could be to explicitly specify the available version numbers to the argument parser via a `choice`, or a version number range via a custom argument `type`. Although the latter would require us validating that our major version numbers are consecutive (e.g. in `kcidb_io.schema.misc.Version.__init__()`). Both of these could wait until we're ready to unify schema version specification.

spbnick reviewed (changes requested):
Thanks, Paweł!

Ah, pardon me, if I was unclear. We don't need to add any positional arguments to the `kcidb.validate_main()` function itself, but rather add the specification for the positional argument to the command-line argument parser created there. I left an inline comment with a suggestion on how to do that.

pawiecz commented:
Thanks Nikolai! I appreciate your detailed explanations. Let me know if v3 needs any adjustments.

spbnick reviewed (changes requested):
Thanks, @pawiecz, looks good now, only one tiny omission to fix.

pawiecz commented on kcidb/__init__.py:
```diff
@@ -250,10 +250,27 @@ def validate_main():
+            version = version.previous
+        if not version:
+            raise Exception(
+                "No version found for major number {args.schema_version}"
```
Marked as "f-string", added a test to validate error message. I intended to test against schema "v0" but even though 0 is accepted by parser (as a `non_negative_int`) it's resolved as falsy in schema version selection and `LATEST` is set.

Testing against schema v999 should buy some time to implement [`choice`/specific `type` for argument parser](https://github.com/kernelci/kcidb/pull/221#discussion_r744621402).

spbnick replied:
Right, we should check against None explicitly, then. I did that, switched to testing with v0 and rebased your commit. Should be ready to go in after CI is done with it.

Thanks a lot, Paweł!
